spring-batch-core/src/main/java/org/springframework/batch/core/repository/JobRepository.java

```diff
@@ -178,14 +178,28 @@ JobExecution createJobExecution(String jobName, JobParameters jobParameters)
 	 */
 	void updateExecutionContext(JobExecution jobExecution);
 
+	/**
+	 * @return all step executions, from all job executions, for the given job instance.
+	 */
+	Collection<StepExecution> getStepExecutions(JobInstance jobInstance);
+
 	/**
 	 * @param jobInstance {@link JobInstance} instance containing the step executions.
 	 * @param stepName the name of the step execution that might have run.
-	 * @return the last execution of step for the given job instance.
+	 * @return the last execution of step for the given job instance, null otherwise.
 	 */
 	@Nullable
 	StepExecution getLastStepExecution(JobInstance jobInstance, String stepName);
 
+	/**
+	 *
+	 * @param stepExecutions all step executions to filter through
+	 * @param stepName the name of the step execution that might have run.
+	 * @return the last execution of step inside the list, null otherwise.
+	 */
+	@Nullable
+	StepExecution getLastStepExecution(Collection<StepExecution> stepExecutions, String stepName);
```

Comment:
Filtering a collection of objects by name is not the job of `JobRepository`. This method could be in a utility class or private/protected to a given implementation, but not part of the `JobRepository` contract. Please remove the method from the interface and update the implementations accordingly.

Here is how I would do it in `SimpleStepExecutionSplitter`:

```java
@Nullable
protected StepExecution getLastStepExecution(Collection<StepExecution> stepExecutions, String stepName) {
   return stepExecutions.stream()
      .filter(stepExecution -> stepExecution.getStepName().equals(stepName))
      .min(new StepExecutionComparator())
      .orElse(null);
}
```

The code of the comparator is what is done here: https://github.com/spring-projects/spring-batch/blob/master/spring-batch-core/src/main/java/org/springframework/batch/core/repository/support/SimpleJobRepository.java#L236-L246